### DISCUSSION TITLE
Helper function to copy URL Or Notes Body in ekg-notes mode

### DISCUSSION THREAD

**Discussion by jayrajput** [ahyatt/ekg#45] [Ideas]
Similar to `b` which can be used to browse the url, new command can be added. I would prefer two new commands to copy either the url or the notes body mapped to `w` and `W`.

Copying URL was easy enough and can be modelled based on `ekg-browse-url`. Copying content is something which I still need to figure out.

The use cases are to copy the URL or content when you wanted to share some parts of your notes with a colleague during chats. I am not sure if embark helps here to change the target before copying. Other than copying URL and body, I am not sure if it will helps to copy/edit tags.

**Comment by qingshuizheng** [comment]
@jayrajput 

The following just works on my end: 

```elisp
;;;###autoload
(defun z/ekg-notes-copy-id-as-kill ()
  "Copy current note id."
  (interactive nil ekg-notes-mode)
  (let* ((note (ekg--current-note-or-error))
         (id (ekg-note-id note)))
    (kill-new (cond ((numberp id) (number-to-string id))
                    ((stringp id) id))))
  (message "Note id saved to kill-ring."))

;;;###autoload
(defun z/ekg-notes-copy-text-as-kill ()
  "Copy current note text, with inlines transcluded."
  (interactive nil ekg-notes-mode)
  (let* ((note (ekg--current-note-or-error)) ; <- note displayed in ewoc
         (id (ekg-note-id note))
         (note (ekg-get-note-with-id id)) ; <- note stored in db (true value)
         (text (ekg-note-text note)))
    (kill-new (ekg-display-note-text note)))
  (message "Note content saved to kill-ring."))

```

**Reply by jayrajput** [reply]
Thanks. I will give them a try.

**Comment by ahyatt** [comment]
I think something like this should be in ekg, but I'm not sure whether the resource, or the text, or what is most reasonable to copy.  Perhaps this is where embark could come in handy.

**Reply by jayrajput** [reply]
Let me try to work on the `embark` and see if I can get that working.

**Reply by jayrajput** [reply]
I did looked a little into embark and I cannot quite figure out how to define a new embark target for ekg notes. I will appreciate any guidance on this.

**Comment by jayrajput** [comment]
@qingshuizheng , @ahyatt 

I was able to figure out how to add the targets to the embark. See code changes below. What do you say?

@ahyatt Shall  `ekg--current-note-or-error` be a public function. This looks like a nice function which the `ekg` user may want to use in their code.

```
    (defun jr/embark-target-ekg-note-url ()
      (when (eq major-mode 'ekg-notes-mode)
	(cons 'ekg-note-url (ekg-note-id (ekg--current-note-or-error)))))

    (defun jr/embark-target-ekg-note-text ()
      (when (eq major-mode 'ekg-notes-mode)
	(cons 'ekg-note-text (ekg-note-text (ekg--current-note-or-error)))))

    (add-to-list 'embark-target-finders #'jr/embark-target-ekg-note-text)
    (add-to-list 'embark-target-finders #'jr/embark-target-ekg-note-url)
```

**Reply by ahyatt** [reply]
I agree, I'll make that change soon, thank you for bringing it up.

**Reply by qingshuizheng** [reply]
> I was able to figure out how to add the targets to the embark

Awesome! Finally for embark users 😃 

I don't use embark on ekg-notes-mode at the moment, since the kbds in ekg-notes-map are good enough for me.

One thing for `ekg-note-text`, if need to render INLINES while copying, then render it with `ekg-display-note-text` is a must:

```elisp
(defun jr/embark-target-ekg-note-text ()
  (when (derived-mode-p 'ekg-notes-mode)
	  (cons 'ekg-note-text
          ;; (ekg-note-text (ekg--current-note-or-error)) ;; before
          (ekg-display-note-text (ekg-get-note-with-id (ekg-note-id (ekg--current-note-or-error)))) ;; after
          )))
```

For `ekg-note-id`, if resource is integer, it should be converted to string with `number to string` before embark-copy-as-kill can be used.

**Reply by jayrajput** [reply]
Thanks for the suggestion. I will give them a try. Still playing with them to see how it goes.

Embark has some nice usage in terms of tag searching. I start with calling `ekg-show-notes-with-any-tags` with a global key chord and then selecting the tags. Lot of time, you want to select more than one tag e.g. tags having the same prefix. In such cases, I invoke `embark-act-all` to invoke `ekg-show-notes-with-any-tags` to act on all. This open up multiple buffer which I review. I can easily kill the buffer using `q` kbd.

I am still working on how to pass all the minibuffer selections as an argument to single ekg function. But I feel the possibilities are endless and there should be someway to do that.

I also like the `embark-collect` to collect tags and then play with them.

**Reply by jayrajput** [reply]
While typing I realize I can just use `embark-become` for my flow to convert `ekg-show-notes-with-any-tags` to a new function which can call `ekg-show-notes-with-any-tags` after passing the current minibuffer input through `ekg-tags-including`.

**Reply by ahyatt** [reply]
I've checked in the change 3b09c8c688533dbce8fb85dbc934488d0c3cf2c5 to `develop`, to promote `ekg--current-note-or-error` and a few other useful functions to be more public module functions.

**Reply by jayrajput** [reply]
Thanks!!
